examples/misc/raptor.py

```diff
@@ -108,128 +144,152 @@ def task_state_cb(task, state):
                                 ]
             tds.append(td)
 
-        pmgr  = rp.PilotManager(session=session)
-        tmgr  = rp.TaskManager(session=session)
+        pmgr = rp.PilotManager(session=session)
+        tmgr = rp.TaskManager(session=session)
+        tmgr.register_callback(task_state_cb)
+
         pilot = pmgr.submit_pilots(pd)
+        tmgr.add_pilots(pilot)
+
+        logger.info('Submit raptor master(s) %s', [t.uid for t in tds])
         task  = tmgr.submit_tasks(tds)
+        if not isinstance(task, list):
+            task = [task]
+
+        tmgr.wait_tasks(uids=[t.uid for t in task], state=rp.AGENT_EXECUTING)
 
       # pmgr.wait_pilots(uid=pilot.uid, state=[rp.PMGR_ACTIVE])
+        logger.info('Stage files for the worker `my_hello` command.')
+        # TODO: Where is this used?
         pilot.stage_in({'source': ru.which('radical-pilot-hello.sh'),
                         'target': 'radical-pilot-hello.sh',
                         'action': rp.TRANSFER})
-        pilot.prepare_env(env_name='ve_raptor',
-                          env_spec={'type'   : 'venv',
-                                    'path'   : '/tmp/ve_raptor/',
-                                    'setup'  : ['$HOME/j/rp/',
-                                                '$HOME/j/ru/',
-                                                'mpi4py']})
 
-        # submit some test tasks
+        # Issue an RPC to provision a Python virtual environment for the later
+        # raptor tasks.  Note that we are telling prepare_env to install
+        # radical.pilot and radical.utils from sdist archives on the local
+        # filesystem. This only works for the default resource, local.localhost.
+        logger.info('Call pilot.prepare_env()')
+        pilot.prepare_env(env_name='ve_raptor',
+                          env_spec={'type' : 'venv',
+                                    'path' : '/tmp/ve_raptor/',
```

Comment:
Is the future environment variable available for expansion?
```suggestion
                                    'path' : '$RP_RESOURCE_SANDBOX/ve_raptor/',
```
If not, can we use a `ru.Url`, like `resource:///ve_raptor/`?

Reply:
Alas no, not supported at the moment :-/

Reply:
note that `path` should be optional. RP should choose something reasonable.

Reply:
```suggestion
```